src/Akka.Streams.Kafka/Helpers/PromiseControl.cs

```diff
@@ -1,65 +1,85 @@
 using System;
 using System.Threading.Tasks;
-using Akka.Streams.Kafka.Extensions;
-using Akka.Streams.Util;
 
-#nullable enable
 namespace Akka.Streams.Kafka.Helpers
 {
+    internal static class PromiseControl
+    {
+        public interface IControlOperation;
+
+        public sealed class ControlStop : IControlOperation
+        {
+            public static readonly ControlStop Instance = new();
+            private ControlStop(){}
+        }
+
+        public sealed class ControlShutdown : IControlOperation
+        {
+            public static readonly ControlShutdown Instance = new();
+            private ControlShutdown(){}
+        }
+    }
+    
     /// <summary>
     /// Used in source logic classes to provide <see cref="IControl"/> implementation.
     /// </summary>
     /// <typeparam name="TSourceOut"></typeparam>
-    internal abstract class PromiseControl<TSourceOut> : IControl
+    internal class PromiseControl<TSourceOut> : IControl
     {
         private readonly SourceShape<TSourceOut> _shape;
         private readonly Action<Outlet<TSourceOut>> _completeStageOutlet;
         private readonly Action<bool> _setStageKeepGoing;
 
-        private readonly TaskCompletionSource<Done> _shutdownTaskSource = new TaskCompletionSource<Done>();
-        private readonly TaskCompletionSource<Done> _stopTaskSource = new TaskCompletionSource<Done>();
-        private readonly Action _stopCallback;
-        private readonly Action<Exception?> _shutdownCallback;
+        private readonly TaskCompletionSource<Done> _shutdownTaskSource = new();
+        private readonly TaskCompletionSource<Done> _stopTaskSource = new();
+        private readonly Action<PromiseControl.IControlOperation> _controlCallback;
+        private readonly Action _performShutdown;
 
         public PromiseControl(
             SourceShape<TSourceOut> shape, 
             Action<Outlet<TSourceOut>> completeStageOutlet, 
             Action<bool> setStageKeepGoing,  
-            Func<Action, Action> asyncCallbackFactory,
-            Func<Action<Exception?>, Action<Exception?>> asyncShutdownCallbackFactory)
+            Func<Action<PromiseControl.IControlOperation>, Action<PromiseControl.IControlOperation>> asyncShutdownCallbackFactory,
+            Action performShutdown)
         {
             _shape = shape;
             _completeStageOutlet = completeStageOutlet;
             _setStageKeepGoing = setStageKeepGoing;
-
-            _stopCallback = asyncCallbackFactory(PerformStop);
-            _shutdownCallback = asyncShutdownCallbackFactory(PerformShutdown);
+            _performShutdown = performShutdown;
+            _controlCallback = asyncShutdownCallbackFactory(c =>
+            {
+                switch (c)
```

Comment:
This is one of three big fixes: these mechanics basically force the correct shutdown method to be invoked properly, which wasn't happening before.